File: behaviour/elixir/gen_mod.ex
## gen_mod.ex
defmodule GenMod do
  @callback do_start   :: :ok
  @callback do_connect :: :ok

  def start(module) do
    IO.puts("[GenMod] Starting Module #{Kernel.inspect(module)}")
    module.do_start
    :ok
  end

  def connect(module) do
    IO.puts("[GenMod] Connecting Module #{Kernel.inspect(module)}")
    module.do_start
    :ok
  end
end

File: behaviour/erlang/gen_mod.erl
%% gen_mod.erl
-module(gen_mod).
-export([start/1, connect/1, info/0]).

-callback do_start()  -> 'ok'.
-callback do_connect() -> 'ok'.
-optional_callbacs([info/0]).

start(Module) ->
  io:format("[erl gen_mod] Starting ~p~n", [Module]),
  apply(Module, do_start, []), 'ok'.

connect(Module) ->
  io:format("[erl gen_mod] Connecting ~p~n", [Module]),
  apply(Module, do_connect, []), 'ok'.

info() ->
  'Default implementation'.


% c("gen_mod.erl").
